xfrd.c

```diff
@@ -2377,77 +2511,106 @@ find_same_master_notify(xfrd_zone_type* zone, int acl_num_nfy)
-					zone->soa_disk_acquired = 0;
-					zone->soa_nsd_acquired = 0;
-				} else {
-					/* revert soa to the one in server */
-					zone->soa_disk_acquired = zone->soa_nsd_acquired;
-					zone->soa_disk = zone->soa_nsd;
-				}
-				/* pretend we are notified with disk soa.
-				   This will cause a refetch of the data, and reload. */
-				xfrd_handle_incoming_notify(zone, &dumped_soa);
-				xfrd_set_timer_refresh(zone);
-			} else if(zone->soa_disk_acquired >= xfrd->reload_cmd_last_sent) {
-				/* this zone still has to be loaded,
-				   make sure reload is set to be sent. */
-				if(xfrd->need_to_send_reload == 0 &&
-					xfrd->reload_added == 0) {
-					log_msg(LOG_ERR, "xfrd: zone %s: needs "
-									 "to be loaded. reload lost? "
-									 "try again", zone->apex_str);
-					xfrd_set_reload_timeout();
-				}
+			xfrd_soa_type soa;
+			soa.serial = htonl(xfr->msg_new_serial);
+			log_msg(LOG_ERR, "xfrd: zone %s: soa serial %u update "
+			                 "failed, restarting transfer "
+			                 "(notified zone)",
+			                 zone->apex_str, xfr->msg_new_serial);
+			/* revert the soa; it has not been acquired properly */
+			if(xfr->acquired == zone->soa_nsd_acquired) {
+				/* this was the same as served,
+				 * perform force_axfr , re-download
+				 * same serial from master */
+				zone->soa_disk_acquired = 0;
+				zone->soa_nsd_acquired = 0;
+			} else {
+				/* revert soa to the one in server */
+				zone->soa_disk_acquired = zone->soa_nsd_acquired;
+				zone->soa_disk = zone->soa_nsd;
+			}
+			/* fabricate soa and trigger notify to refetch and
+			 * reload update */
+			memset(&soa, 0, sizeof(soa));
+			soa.serial = htonl(xfr->msg_new_serial);
+			xfrd_handle_incoming_notify(zone, &soa);
+			xfrd_set_timer_refresh(zone);
+			/* delete all pending updates */
+			xfr = zone->latest_xfr;
+			while(xfr) {
+				if(!xfr->acquired)
+					continue;
+				prev_xfr = xfr->prev;
+				xfrd_delete_zone_xfr(zone, xfr);
+				xfr = prev_xfr;
 			}
 		}
 	}
 }
 
 void
-xfrd_prepare_zones_for_reload()
+xfrd_prepare_zones_for_reload(void)
 {
 	xfrd_zone_type* zone;
+	xfrd_xfr_type* xfr;
+	int reload, send;
+
+	send = 1;
+	reload = 0;
 	RBTREE_FOR(zone, xfrd_zone_type*, xfrd->zones)
 	{
-		/* zone has a disk soa, and no nsd soa or a different nsd soa */
-		if(zone->soa_disk_acquired != 0 &&
-			(zone->soa_nsd_acquired == 0 ||
-			zone->soa_disk.serial != zone->soa_nsd.serial))
-		{
-			if(zone->soa_disk_acquired == xfrd_time()) {
-				/* antedate by one second.
-				 * this makes sure that the zone time is before
-				 * reload, so that check_failed_zones() is
-				 * certain of the result.
-				 */
-				zone->soa_disk_acquired--;
+		xfr = zone->latest_xfr;
+		while(xfr) {
+			if(!xfr->prev)
+				break;
+			xfr = xfr->prev;
+			assert(xfr->acquired);
+		}
+
+		while(xfr && xfr->acquired) {
+			/* skip updates that arrived after failed reload */
+			if(xfrd->reload_cmd_first_sent && !xfr->sent)
+				continue;
```

Comment:
This is also an infinite loop if the condition is true, if !xfr->sent, it should go to xfr->next before continue.

Reply:
done, should've been a `break`